Settings for notifications

alkari-verende opened the pull request:
Several users have requested the ability to reduce the number of notifications sent from the calendar. This commit adds two new setting groups:

First, the ability to globally disable each type of notification: create/post, update/edit, cancel, activate, end. For us, we find the "update" notifications more noisy than helpful, so we disable it as shown in the screenshot.

Second, this adds the ability to change the interval on which SeAT will ping out reminders for operations:

<img width="442" alt="Screen Shot 2021-12-31 at 9 15 35 PM" src="https://user-images.githubusercontent.com/70491080/147842698-be1a1a58-fd62-431d-8482-0cff85e96c9f.png">

Note that these settings are global to all events.

Closes #22 
Closes #23

warlof commented on src/Commands/RemindOperation.php:
```diff
@@ -25,24 +25,28 @@ class RemindOperation extends Command
+        foreach ($marks as $mark)
+        {
+            $when = Carbon::now('UTC')->floorMinute()->addMinutes($mark);
+            $ops = Operation::where('is_cancelled', 'false')
```
this must be a `boolean`

warlof commented on src/Http/Controllers/SettingController.php:
```diff
@@ -50,6 +51,36 @@ public function updateSlack(Request $request)
             $validated_data['slack_integration_default_channel'],
         ], true);
 
+        setting([
```
At this stage, it would be better to spawn a validation class and update the method signature.
You'll be able to check all form inputs, types and still beeing able to call value like an object.
- https://laravel.com/docs/6.x/validation#form-request-validation

You'll find implementation sample in core :
- https://github.com/eveseat/web/blob/master/src/Http/Validation/SeatSettings.php
- https://github.com/eveseat/web/blob/master/src/Http/Controllers/Configuration/SeatController.php#L86-L94

That will also help you to reduce method complexity.

warlof commented on src/Observers/OperationObserver.php:
```diff
@@ -34,15 +36,17 @@ public function updating(Operation $new_operation)
         if (setting('kassie.calendar.slack_integration', true) == 1 && !is_null($new_operation->integration)) {
             $old_operation = Operation::find($new_operation->id);
             if ($old_operation->is_cancelled != $new_operation->is_cancelled) {
```
will not a `switch` be better here ?

alkari-verende replied:
I don't believe so, because the conditional logic is different for each clause. I'm not sure how a switch statement would make this clearer?  For whatever it's worth, this is not new logic, I just added the check for notification settings.

warlof commented on src/database/migrations/2021_12_31_213410_default_notification_settings.php:
```diff
@@ -0,0 +1,39 @@
+class DefaultNotificationSettings extends Migration
+{
+    const SETTINGS = [
+        'kassie.calendar.notify_create_operation',
```
even better, using an array key when key is setting `$name` and value is setting `$value`.
This way, you can init the new frequency field with default `5,15,30`.

warlof commented on src/database/migrations/2021_12_31_213435_default_ping_operation_settings.php:
```diff
@@ -0,0 +1,31 @@
+<?php
```
I think you can merge both migration here

warlof commented on src/Commands/RemindOperation.php:
```diff
@@ -25,24 +25,28 @@ class RemindOperation extends Command
+        # Ensure we send reminders starting with furthest in the future. That way
+        # when more than one event is being reminded, the last reminder in chat
+        # is the next event to occur.
+        $configured_marks = setting('kassie.calendar.notify_operation_interval', true);
```
will not this change also require an update of the schedule seeder -> https://github.com/BenHUET/eveseat-calendar/blob/master/src/database/seeds/ScheduleSeeder.php#L15 ?

Currently, it's not designed to be called every minute.

alkari-verende replied:
From what I can see, the schedule seeder is not in use. The reminder task is being executed from cron, configured in the service provider: https://github.com/BenHUET/eveseat-calendar/blob/master/src/CalendarServiceProvider.php#L29

Maybe have a separate PR to either migrate complete to the schedule seeder, or remove it?

warlof replied:
hum, indead - not address by this pull request
you had a nice catch with the provider - that also mean currently, the plugin is calling the command by default - so peoples may have duplicate calls if they also put it into their schedulers config.

Logic would be to use the seeder instead of the provider override - if you got a bit of time to make that switch and update the seeder accordingly ?

alkari-verende replied:
This PR is already sizeable, and I would like to keep it focused on the changes specifically around notification settings. I am happy to continue the work to update the seeder but I would prefer to do that as a separate PR, and get this one merged sooner.

warlof commented on src/resources/lang/en/seat.php:
```diff
@@ -107,15 +107,25 @@
-    'save'              => 'Save',
+    'notifications_to_send' => 'Notifications to send',
+
+    'help_notify_operation_interval' => 'Decide how many pings to send before each operation. Each value is the number of minutes prior to the operation to send the ping. Separate numbers with commas. Default value of "15,30,60" will send 3 notifications: 15 minutes, 30 minutes, and 60 minutes prior to the operation start time.',
```
well done, maybe just add the value in a `code` tag so it's be more visible and put it as a translation parameter inside the view.
- https://laravel.com/docs/8.x/localization#replacing-parameters-in-translation-strings

alkari-verende replied:
<img width="405" alt="Screen Shot 2022-01-01 at 12 26 50 PM" src="https://user-images.githubusercontent.com/70491080/147856382-2a0601d2-9735-41a7-a6e0-d44839a38da0.png">

alkari-verende commented:
@warlof Thank you again for all the feedback. I believe I have addressed all the comments, except the two I responded to above. Can I get another review?
